### PR TITLE
[WIP] Popup documentation

### PR DESCRIPTION
**Pull request by thomasfaingnaert**
Uses the hover logic to display documentation in the popup/floating window (like `completeopt+=popup`, but with Markdown capabilities).

Vim:
vim

Neovim:
nvim

TODO:
- [x] Cleanup Neovim window
- [x] Neovim positioning
- [ ] Use `completeopt+=popup` positioning logic for Vim (?)
- [ ] Allow disabling

### CODE REVIEW

**Comment by thomasfaingnaert** [comment]
Hmm, that's fired on BufWinLeave. Might be worth it to log `expand('<afile>')` to see which buffer is being removed?

**Comment by clason** [comment]
OK, with a bit more careful bisecting, I tracked it down to https://github.com/neovim/neovim/pull/9807/files#diff-40d9eb21ee75b3df148c827ec4d4310aR4264, specifically line 4264 in `src/nvim/window.c`. Uncommenting this line on master removes the spurious calls to `get_matching_delimiters`. I still don't understand why you can't reproduce my issue, even though that change should be in your version.

**Comment by thomasfaingnaert** [comment]
Wiped my config, and tried again, still no luck.

Neovim version:
```
NVIM v0.4.0-2056-g11fe132dd
Build type: RelWithDebInfo
LuaJIT 2.0.5
Compilation: /usr/bin/cc -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -DMIN_LOG_LEVEL=3 -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=always -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -I/home/thomas/src/neovim/build/config -I/home/thomas/src/neovim/src -I/home/thomas/src/neovim/.deps/usr/include -I/usr/include -I/home/thomas/src/neovim/build/src/nvim/auto -I/home/thomas/src/neovim/build/include
Compiled by thomas@dell-xps

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/local/share/nvim"

Run :checkhealth for more info
```

init.vim
```vim
call plug#begin()
Plug 'lervag/vimtex'
Plug 'prabirshrestha/async.vim'
Plug '~/projects/vim-lsp'
call plug#end()

autocmd User lsp_setup call lsp#register_server({
			\ 'name': 'texlab',
			\ 'cmd': {server_info->['texlab']},
			\ 'whitelist': ['tex'],
			\ })
autocmd FileType tex setlocal omnifunc=lsp#complete

set completeopt+=noinsert

let g:lsp_log_verbose = 1
let g:lsp_log_file = expand('~/vim-lsp.log')
```

vimtex patch
```diff
diff --git a/autoload/vimtex/delim.vim b/autoload/vimtex/delim.vim
index f56dac6..e991de1 100644
--- a/autoload/vimtex/delim.vim
+++ b/autoload/vimtex/delim.vim
@@ -807,6 +807,7 @@ endfunction
 
 " }}}1
 function! s:get_matching_delim() dict abort " {{{1
+  call lsp#log('vimtex delim')
   let [re, flags, stopline] = self.is_open
         \ ? [self.re.close,  'nW', line('.') + s:stopline]
         \ : [self.re.open,  'bnW', max([line('.') - s:stopline, 1])]
```

test.tex
```tex
\documentclass[a4paper,11pt]{article}

\begin{document}

\section{a}
\label{sec:a}

\section{b}
\label{sec:b}

\section{c}
\label{sec:c}

\cref{

\end{document}
```

Cursor is at the end of cref line (delim is called).
I enter insert mode with `a`, it's called again.
Entering completion with `<C-x><C-o>`, pressing `<up>`/`<down>`, or even `<C-n>` and `<C-p>` doesn't trigger it anymore.

Am I doing something wrong?

**Comment by clason** [comment]
Well, different: My test example has `equation` environments -- if you replace the `sections` by
```latex
\begin{equation}\label{eq1}
  1+1=2
\end{equation}
```
you should see the behavior.

**Comment by thomasfaingnaert** [comment]
As an aside, I played a bit more with popups, and I managed to get `completionItem/resolve` and markdown working with typescript language server. It uses the popup Vim creates with `completeopt+=popup`, so you can align it to items, and Vim handles the sizing itself:

Peek 2019-09-13 15-27

**Comment by clason** [comment]
Currently discussing this on https://gitter.im/neovim/neovim, by the way; I think the output routines for neovim should be refactored to avoid this issue.

**Comment by thomasfaingnaert** [comment]
@clason Have been following the gitter somewhat. I guess the conclusion is that `output.vim` should be refactored to not enter the floating window?

**Comment by clason** [comment]
Yes, and I'm working on it. Only thing that's missing is to have the float resize to match the contents; can't get the actual number of lines of preview text to do that.

**Comment by clason** [comment]
How do I get the maximum width needed to show `a:lines`?

**Comment by thomasfaingnaert** [comment]
Is this what you're looking for? https://github.com/prabirshrestha/vim-lsp/blob/da078b1096944d149a0593dc8e49ae097d7b4ed3/autoload/lsp/ui/vim/output.vim#L323-L337

**Comment by thomasfaingnaert** [comment]
While we're on the topic of sizing; do you think the Neovim devs would consider handling the sizing in Neovim itself, so plugins wouldn't need to worry about it? The sizing API of Vim popups is much nicer in that regard; you just set the content, minwidth, maxwidth, minheight, maxheight and position, and Vim handles the rest.

**Comment by clason** [comment]
They seem to be quite open for suggestions, especially for APIs (and don't mind following vim API where it makes sense).

**Comment by thomasfaingnaert** [comment]
Do you have time to create an issue for this?

**Comment by clason** [comment]
Actually, it might be better for you to do it since you know much better what vim-lsp would need; I could open an issue, but would have to go back-and-forth with you about it for clarification. 

I could ask on gitter for you first, though?

**Comment by clason** [comment]
I think I have it almost working now, but for two issues:
1. Preview floats contents are right-aligned (meaning the start is cut off rather than the end if the contents don't fit the screen)
2. It doesn't show if triggered on the first line of the buffer.

**Comment by thomasfaingnaert** [comment]
Sounds good; the thing is that I'm not really familiar with nvim's API. Most of my knowledge comes from just reading `output.vim`. One thing that immediately catched my eye was that `s:adjust_float_placement` is only used for Neovim, so that's why I'm guessing that in Neovim you have to position and size the floats yourself. 
Anyway, mirroring Vim's `popup_create-arguments` would be nice (if that's doable at least).

**Comment by prabirshrestha** [comment]
Not required for this but how would function overload pop up work?

**Comment by thomasfaingnaert** [comment]
@prabirshrestha 
You love making thinks more complicated, don't you :p 
I've made an issue at Vim to expand the API for info popups at https://github.com/vim/vim/issues/4924, could you take a look at it and share your thoughts?
With the extra function `complete_set_popup`, overloads would be possible too, I suppose, provided we have a way to catch keypresses

**Comment by prabirshrestha** [comment]
vim-lsp is already taken so can't create it.

I created one in discord instead. Here is the invitation link that lasts for 1 day. https://discord.gg/zpvutT

**Comment by clason** [comment]
@prabirshrestha I think you can ask github to transfer the organization since the current owner is not doing anything with it -- if you want.

**Comment by prabirshrestha** [comment]
@clason We had asked that before, they do have private repos that are active.

**Comment by clason** [comment]
Ah, OK, that's all right then. Thanks for letting me know!

**Comment by thomasfaingnaert** [comment]
Do you need an organisation for gitter? I think you can create `prabirshrestha/vim-lsp` as well.

**Comment by thomasfaingnaert** [comment]
@prabirshrestha Any update on gitter?

**Comment by clason** [comment]
@thomasfaingnaert Yes, that should work -- the (short-lived) chatroom I mentioned was called `clason/vim-lsp` and connected to my fork.

**Comment by prabirshrestha** [comment]
I actually own this org https://github.com/vimlsp. The one without dash. So finally was able to create this. https://gitter.im/vimlsp/community.

Let me know if you want to create specific rooms. In the meantime I created https://gitter.im/vimlsp/setup for those that want to ask how to setup servers.

**Comment by prabirshrestha** [comment]
@thomasfaingnaert any updates on this?

**Comment by prabirshrestha** [comment]
@thomasfaingnaert any updated on this? Would be also good to fix merge conflicts.

**Comment by thomasfaingnaert** [comment]
@prabirshrestha I haven't had the time to work on this, unfortunately. I'll try to pick this up again in the summer.

File: autoload/lsp/ui/vim/documentation.vim [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        if s:use_nvim_float | call nvim_win_close(s:last_popup_id, 1) | endif
+
+        let s:last_popup_id = -1
+    endif
+endfunction
+
+function! lsp#ui#vim#documentation#setup() abort
+    augroup lsp_documentation_popup
+        autocmd!
+        autocmd CompleteChanged * call s:complete_done()
```
Comment (prabirshrestha) [inline autoload/lsp/ui/vim/documentation.vim:72]: we do need a flag to enable or disable the popup docs.
We need to also document this.

**Comment by prabirshrestha** [comment]
@thomasfaingnaert do you have the branch with `completionItem/resolve` support that you pushed? Would be good to have a look into it too.

**Comment by prabirshrestha** [comment]
In the meantime `doc-popup` branch contains the merge conflict fixes.

**Comment by thomasfaingnaert** [comment]
> @thomasfaingnaert do you have the branch with `completionItem/resolve` support that you pushed? Would be good to have a look into it too.

I don't think I have that branch still around, unfortunately. IIRC, there was not much to it: store the completion items in a list, and update the current completion item in that list on `CompleteChanged`.

File: autoload/lsp/ui/vim/documentation.vim (hunk 1 above)
Comment (prabirshrestha) [reply autoload/lsp/ui/vim/documentation.vim:73]: fixed.

**Comment by prabirshrestha** [comment]
@thomasfaingnaert no worries. Just wanted to know if there was support so I can review. For now will review the current one with latest fix and merge and can support `completionItem/resolve` later. Ideally it would had been good if neovim supported `completeopt+=popup`. This would had reduce lot of our logic.

Thanks for this PR.

**Pull request merged**